rfcs/0046-Intel-LIBXSMM-integration.md

```diff
@@ -0,0 +1,108 @@
+```python
+  from tvm.relay.op.contrib import libxsmm
+
+  # API to call LIBXSMM partitioning
+  libxsmm_module = libxsmm.partition_for_libxsmm(module) 
+```
+  * Pattern matching table: 
+```python
+  @register_pattern_table("libxsmm")
+  def pattern_table():
+      dense_pattern = ("libxsmm.dense", make_pattern(with_bias=False, with_activation=None))
+      denese_bias_pattern = ("libxsmm.dense_bias", make_pattern(with_bias=True, with_activation=None))
+      denese_relu_pattern = ("libxsmm.dense_relu", make_pattern(with_bias=False, with_activation="relu"))
+      denese_sigmoid_pattern = ("libxsmm.dense_sigmoid", make_pattern(with_bias=False, with_activation="sigmoid"))
+      denese_bias_relu = ("libxsmm.dense_bias_relu", make_pattern(with_bias=True, with_activation="relu"))
+      denese_bias_sigmoid = ("libxsmm.dense_bias_sigmoid", make_pattern(with_bias=True, with_activation="sigmoid"))
+      libxsmm_pattern = [dense_pattern, denese_bias_pattern, denese_relu_pattern, denese_sigmoid_pattern, denese_bias_relu, denese_bias_sigmoid]
+      return libxsmm_pattern
+```
+  * Build with TVM
+```python
+  with tvm.transform.PassContext(opt_level=3):
+    lib = relay.build(libxsmm_module, target="cpu", params=params)
+```
+3. Integrate into TOPI, an GEMM autotvm template with LIBXSMM as inner kernel.
+  * Use Tensorize/TensorIR to substitute register block of GEMM with LIBXSMM
+```python
+  def intrin_func(ins, outs):
+    def _body():
+      ib = tvm.tir.ir_builder.create()
+      ib.emit(
+        tvm.tir.call_extern(
+          "int", "libxsmm_sgemm", m, n, k, 1.0, ins[0].access_ptr("r"), K, ins[1].access_ptr("r"), n, 0.0, outs[0].access_ptr("w"), N
+        )
+      )
+      return ib.get()
+
+    def _update():
+      ib = tvm.tir.ir_builder.create()
+      ib.emit(
+        tvm.tir.call_extern(
+           "int", "libxsmm_sgemm", m, n, k, 1.0, ins[0].access_ptr("r"), K, ins[1].access_ptr("r"), n, 1.0, outs[0].access_ptr("w"), N
+        )
+      )
+      return ib.get()
+```
+
+# Testing
+We will add unittest for coresponding extern call, BYOC and TOPI related code:
+* Make sure the result LIBXSMM produces is correct with its TVM counter part;
+* Confirm match patterns are working as expected.
+
+# Drawbacks
+* Though LIBXSMM works well with AutoTVM, it does not help AutoScheduler;
+* Memory footprint would increase as JIT code generated, a LRU kernel cache might be required to mitigate it. 
+
+# Future possibilities
+* LIBXSMM has DNN support, so it might be interesting to also integrate DNN primitives such as Conv to TVM;
+* LIBXSMM has quantized kernel (int8), we can also integrate it to TVM, as long as it surpass existing oneDNN implementations.
+
+# Upstream plan
+This proposal would be split to following PR series:
+1. Add LIBXSMM as extern call;
+2. Add LIBXSMM to BYOC for accelerating small gemm;
+3. Add LIBXSMM-enabled normal size GEMM to TOPI;
+4. Add LIBXSMM-enabled normal size GEMM to Relay op strategy.
```

Comment:
I reconsidered this plan again and felt that the step 1, 3, 4 are a group while step 2 is relatively independent. After all, you don't need extern ops to make BYOC work. Thus, I would suggest either (1, 3, 4, 2) or (2, 1, 3, 4).

In addition, what I meant by "TVM CI" was that you will need to add LIBXSMM to the docker environment before merging any changes; otherwise your unit tests won't pass the CI. In summary, I would suggest the following PRs:

1. Add LIBXSMM to TVM CI.
2. BYOC support.
3. Documentation about LIBXSMM support, including supported ops/patterns/dtypes/versions and limitations. Ref: https://tvm.apache.org/docs/how_to/deploy/tensorrt.html.
3. Relay/TOPI op support (this can be at step 2 if you prefer).

Reply:
It does make sense to have the order you suggested. I will update the plan ASAP.